acceptance/tests/direct_puppet/static_catalog_env_control.rb

```diff
@@ -0,0 +1,348 @@
+  }
+}',
+}
+MANIFESTAGENT
+
+  step "Add versioned-code parameters to puppetserver.conf and ensure the server is running" 
+  puppetserver_config = "#{master['puppetserver-confdir']}/puppetserver.conf"
+  on master, "cp #{puppetserver_config} #{@coderoot}/puppetserver.conf.bak"
+  versioned_code_settings = {
+    "versioned-code" => {
+      "code-id-command" => "#{@coderoot}/code_id.sh",
+       "code-content-command" => "#{@coderoot}/code_content.sh"
+    }
+  }
+  modify_tk_config(master, puppetserver_config, versioned_code_settings)
+
+    production_file = "#{resdir}/prod_hello_from_puppet_uri"
+    canary_file = "#{resdir}/can_hello_from_puppet_uri"
+    catalog_file = @catalog_files[agent.hostname]
+
+    step 'agent gets a production catalog, should be static catalog by default'
+    on(
+      agent,
+      puppet(
+        'agent',
+        '-t',
+        '--environment', 'production',
+        '--server', master.hostname
+      ),
+      :acceptable_exit_codes => [0, 2]
+    )
+  
+    step 'verify production environment'
+    r = on(agent, "cat #{catalog_file}")
+    catalog_content = JSON.parse(r.stdout)
+    assert_equal(
+      catalog_content['environment'],
+      'production',
+      'catalog for unexpectected environment'
+    )
+
+    step 'verify static catalog by finding metadata section in catalog'
+    assert(
+      catalog_content['metadata'] && catalog_content['metadata'][production_file],
+      'metadata section of catalog not found'
+    )
+
+    step 'agent gets a canary catalog, static catalog should be disabled'
+    on(
+      agent,
+      puppet(
+        'agent',
+        '-t',
+        '--environment', 'canary',
+        '--server', master.hostname
+      ),
+      :acceptable_exit_codes => [0, 2]
+    )
+ 
+    step 'verify canary environment'
+    r = on(agent, "cat #{catalog_file}")
+    catalog_content = JSON.parse(r.stdout)
+    assert_equal(
+      catalog_content['environment'],
+      'canary',
+      'catalog for unexpectected environment'
+    )
+
+    step 'verify not static catalog by absence of metadata section in catalog'
+    assert_nil(
+      catalog_content['metadata'],
+      'unexpected metadata section found in catalog'
+    )
+
+  end 
+end
+
+step 'enable static catalog for canary environment'
+@static_canary_manifest = <<MANIFEST2
+file { '#{@coderoot}/environments/canary/environment.conf':
+  ensure => file,
+  mode => "0644",
+  content => 'environment_timeout = 0
+static_catalogs = true
+',
+}
+MANIFEST2
+apply_manifest_on(
+  master,
+  @static_canary_manifest,
+  :catch_failures => true
+)
+
+step 'disable global static catalog setting'
+@master_opts = {
+  'master' => {
+    'static_catalogs' => false
+  },
+  'main' => {
+    'environmentpath' => "#{@coderoot}/environments",
+  },
+}
+
+step 'bounce server for static catalog disable setting to take effect.'
+with_puppet_running_on master, @master_opts, @coderoot do
+  agents.each do |agent|
```

Comment:
I think this section needs to reapply the init.pp files on the master. I'd pull that section into a helper function.